### DISCUSSION TITLE
merge / alter already created pdf:s

### DISCUSSION THREAD

**Discussion by screwyourshelf** [QuestPDF/QuestPDF#893] [Ideas]
currently using questpdf for creation of pdf files. works well !

in addition to creating the files we need to process the files, later on in the process. We need for instance merge all (100/1000s of files) into a single pdf and "stamp/mark" all the front pages with a certain batch number before we send it off to a 3rd party for distribution etc.

currently we do this post processing with itextaharp but it would be cool to be able to use questpdf also for this.

**Comment by girlpunk** [comment]
Unfortunately QuestPDF doesn't offer any functionality for editing or otherwise opening existing PDFs.

**Comment by MarcinZiabek** [comment]
I am currently preparing a long-term roadmap for the library. I am excited to include this enhancement 😄 

Most likely, the work will be done by using already existing C++ libraries and surfacing their API to the managed C# code. Depending on the complexity and value-added, it will be done as part of QuestPDF or a secondary MIT library.

**Reply by screwyourshelf** [reply]
ah thats cool, any progress on that festure?

**Reply by MarcinZiabek** [reply]
It's already available 😄 

https://www.questpdf.com/concepts/document-operations.html

**Comment by RobertK66** [comment]
I solved my use case with the current document-operations. One could think of some improvements here.

Use Case: I do generate a 'big report' (ca. 1000 pages)  from database and there are some sections where I have external PDFs (ca. 100) to be inserted. My solution does insert empty pages in the first run on the report document and does keep a list of page numbers where to include the other pdfs as overlays after the big one got rendered/generated.

possible(?) Improvements:
- I did not find a way to get the number of pages of the external PDFs with the QuestPDF APIs! Did I miss this somewhere? I used PDFSharp only for retrieving this number in my first pass ....
- The convenient feature here would be to have a layout element "Pdf" which can be filled with file path, byte[] or stream prior to rendering.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. We apologize for the delayed response. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Regarding reading the page count of an external PDF: this is not currently supported, but we are actively researching ways to provide better access to page count information and it is in our backlog. For now, that specific step still requires a dedicated library.

Should you encounter any further problems, please don't hesitate to open a new issue.
Thank you!
